File: memit/amend_check.py
"""
amend_check.py: Triangle inequality check for safe amend detection.
"""
from typing import Dict, Optional, Tuple
from .diff_engine import efficient_edit_distance


def check_amend_safe(
    A_files: Dict[str, Optional[str]],
    B_files: Dict[str, Optional[str]],
    C_files: Dict[str, Optional[str]]
) -> Tuple[bool, str]:
    """
    Check if amending B with C is safe (no information loss).

    Uses triangle inequality: d(A,B) + d(B,C) == d(A,C)
    This means B is on the shortest edit path from A to C.

    Args:
        A_files: Files in snapshot S[-2] (grandparent)
        B_files: Files in snapshot S[-1] (current last snapshot)
        C_files: Files in working directory

    Returns:
        Tuple of (is_safe, reason)
        - is_safe: True if amend is safe
        - reason: Human-readable explanation
    """
    # Collect all file paths across all three versions
    all_files = set(A_files.keys()) | set(B_files.keys()) | set(C_files.keys())

    for file_path in all_files:
        # Get content for each version (treat missing files as empty string)
        a_content = A_files.get(file_path, "")
        b_content = B_files.get(file_path, "")
        c_content = C_files.get(file_path, "")

        # Handle binary files (None content)
        if a_content is None or b_content is None or c_content is None:
            # Binary files: be conservative
            # If binary content changed, don't amend
            if a_content != b_content or b_content != c_content:
                return False, f"Binary file changed: {file_path}"
            # If binary file unchanged across all versions, it's safe
            continue

        # Skip if all three versions are identical
        if a_content == b_content == c_content:
            continue

        # Calculate edit distances
        d_AB = efficient_edit_distance(a_content, b_content)
        d_BC = efficient_edit_distance(b_content, c_content)
        d_AC = efficient_edit_distance(a_content, c_content)

        # If any distance calculation failed (too large), be conservative
        if d_AB is None or d_BC is None or d_AC is None:
            return False, f"File too large for diff: {file_path}"

        # Check triangle inequality: d(A,B) + d(B,C) == d(A,C)
        if d_AB + d_BC != d_AC:
            return False, (
                f"Information loss detected in {file_path}: "
                f"d(A,B)={d_AB} + d(B,C)={d_BC} != d(A,C)={d_AC}"
            )

    # All files passed the triangle inequality test
    return True, "All changes are on the shortest edit path"


def should_amend(
    A_files: Dict[str, Optional[str]],
    B_files: Dict[str, Optional[str]],
    C_files: Dict[str, Optional[str]]
) -> bool:
    """
    Simplified check for whether to amend.

    Args:
        A_files: Files in snapshot S[-2]
        B_files: Files in snapshot S[-1]
        C_files: Files in working directory

    Returns:
        True if amend should be performed
    """
    is_safe, _ = check_amend_safe(A_files, B_files, C_files)
    return is_safe

File: pwa/src/lib/amendCheck.ts
/**
 * amendCheck.ts — Port of amend_check.py
 *
 * 삼각 부등식을 이용해 amend가 안전한지 판단한다.
 * d(A,B) + d(B,C) == d(A,C) 이면 B는 A→C 최단 경로 위에 있으므로 안전.
 */

import { efficientEditDistance } from './diffEngine';

type FileMap = Record<string, string | null>;

export interface AmendResult {
  isSafe: boolean;
  reason: string;
}

export function checkAmendSafe(
  aFiles: FileMap,
  bFiles: FileMap,
  cFiles: FileMap
): AmendResult {
  const allKeys = new Set([
    ...Object.keys(aFiles),
    ...Object.keys(bFiles),
    ...Object.keys(cFiles),
  ]);

  for (const key of allKeys) {
    const a = aFiles[key] ?? '';
    const b = bFiles[key] ?? '';
    const c = cFiles[key] ?? '';

    // null = 바이너리 파일 취급: 변경이 있으면 보수적으로 거부
    if (aFiles[key] === null || bFiles[key] === null || cFiles[key] === null) {
      if (a !== b || b !== c) {
        return { isSafe: false, reason: `Binary file changed: ${key}` };
      }
      continue;
    }

    if (a === b && b === c) continue;

    const dAB = efficientEditDistance(a, b);
    const dBC = efficientEditDistance(b, c);
    const dAC = efficientEditDistance(a, c);

    if (dAB === null || dBC === null || dAC === null) {
      return { isSafe: false, reason: `File too large for diff: ${key}` };
    }

    if (dAB + dBC !== dAC) {
      return {
        isSafe: false,
        reason: `Information loss in ${key}: d(A,B)=${dAB} + d(B,C)=${dBC} != d(A,C)=${dAC}`,
      };
    }
  }

  return { isSafe: true, reason: 'All changes are on the shortest edit path' };
}
